crates/goose/src/message.rs

```diff
@@ -257,13 +258,23 @@ impl MessageContent {
 
 impl From<Content> for MessageContent {
     fn from(content: Content) -> Self {
-        match content {
-            Content::Text(text) => MessageContent::Text(text),
-            Content::Image(image) => MessageContent::Image(image),
-            Content::Resource(resource) => MessageContent::Text(TextContent {
-                text: resource.get_text(),
-                annotations: None,
-            }),
+        match content.raw {
+            RawContent::Text(text) => {
+                MessageContent::Text(text.optional_annotate(content.annotations))
+            }
+            RawContent::Image(image) => {
+                MessageContent::Image(image.optional_annotate(content.annotations))
+            }
+            RawContent::Resource(resource) => {
+                let text = match &resource.resource {
+                    ResourceContents::TextResourceContents { text, .. } => text.clone(),
+                    ResourceContents::BlobResourceContents { blob, .. } => {
+                        format!("[Binary content: {}]", blob.clone())
+                    }
+                };
+                MessageContent::text(text)
+            }
+            RawContent::Audio(_) => todo!(),
```

Comment:
I need to look into how we'd handle Audio--we obviously don't want to panic here